core/src/main/java/org/apache/iceberg/PartitionStats.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+public class PartitionStats implements StructLike {
+
+  private static final int STATS_COUNT = 12;
+
+  private PartitionData partition;
+  private int specId;
+  private long dataRecordCount;
+  private int dataFileCount;
+  private long totalDataFileSizeInBytes;
+  private long positionDeleteRecordCount;
+  private int positionDeleteFileCount;
+  private long equalityDeleteRecordCount;
+  private int equalityDeleteFileCount;
+  private long totalRecordCount;
+  private Long lastUpdatedAt; // null by default
+  private Long lastUpdatedSnapshotId; // null by default
+
+  public PartitionStats(PartitionData partition, int specId) {
+    this.partition = partition;
+    this.specId = specId;
+  }
+
+  public PartitionData partition() {
+    return partition;
+  }
+
+  public int specId() {
+    return specId;
+  }
+
+  public long dataRecordCount() {
+    return dataRecordCount;
+  }
+
+  public int dataFileCount() {
+    return dataFileCount;
+  }
+
+  public long totalDataFileSizeInBytes() {
+    return totalDataFileSizeInBytes;
+  }
+
+  public long positionDeleteRecordCount() {
+    return positionDeleteRecordCount;
+  }
+
+  public int positionDeleteFileCount() {
+    return positionDeleteFileCount;
+  }
+
+  public long equalityDeleteRecordCount() {
+    return equalityDeleteRecordCount;
+  }
+
+  public int equalityDeleteFileCount() {
+    return equalityDeleteFileCount;
+  }
+
+  public long totalRecordCount() {
+    return totalRecordCount;
+  }
+
+  public Long lastUpdatedAt() {
+    return lastUpdatedAt;
+  }
+
+  public Long lastUpdatedSnapshotId() {
+    return lastUpdatedSnapshotId;
+  }
+
+  /**
+   * Updates the partition stats from the data/delete file.
+   *
+   * @param file the ContentFile from the manifest entry.
+   * @param snapshot the snapshot corresponding to the live entry.
+   */
+  public void liveEntry(ContentFile<?> file, Snapshot snapshot) {
+    this.specId = Math.max(this.specId, file.specId());
```

Comment:
This seems a bit suspicious. What about a simple sanity check? I think if spec IDs don't match, it means something went wrong as `PartitionStats` is initialized with the spec ID.

Would something like this be sufficient?

```
Preconditions.checkArgument(specId == file.specId(), "Spec IDs must match");
```

Reply:
this works for `liveEntry`. But we need to updated spec id `appendStats`. Please check below comment.